cmd/util/ledger/migrations/staged_contracts_migration.go

```diff
@@ -148,11 +153,131 @@ func (m *StagedContractsMigration) InitMigration(
 	return nil
 }
 
+// collectAndRegisterStagedContractsFromPayloads scan through the payloads and collects the contracts
+// staged through the `MigrationContractStaging` contract.
+func (m *StagedContractsMigration) collectAndRegisterStagedContractsFromPayloads(allPayloads []*ledger.Payload) error {
+
+	// If the contracts are already passed as an input to the migration
+	// then no need to scan the storage.
+	// TODO: Maybe do this anyway?
+	if len(m.contractsByLocation) > 0 {
+		return nil
+	}
+
+	var stagingAccount string
+
+	switch m.chainID {
+	case flow.Testnet:
+		stagingAccount = "0x2ceae959ed1a7e7a"
+	case flow.Mainnet:
+		stagingAccount = "0x56100d46aa9b0212"
+	default:
+		// For other networks such as emulator etc. no need to scan for staged contracts.
+		return nil
+	}
+
+	stagingAccountAddress := common.Address(flow.HexToAddress(stagingAccount))
+
+	mr, err := NewMigratorRuntime(
+		stagingAccountAddress,
+		allPayloads,
```

Comment:
Creating a payload snapshot for all payloads is prohibitively expensive in some cases (e.g. MN). 

Maybe filter the payloads to just the ones where the owner is `stagingAccountAddress`, given that is all we ever need.

Reply:
good idea, added in https://github.com/onflow/flow-go/pull/5659/commits/1840d292f06360f410f86523e4d639a6b0cc7082